### PR TITLE
Improve FindPodVolumes performance

### PR DESCRIPTION
**Pull request by cofyc**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

It includes three optimizations:

- Merge UpdateProvisionedPVCs with UpdateBindings: https://github.com/kubernetes/kubernetes/pull/72953/commits/7fe97886a82a4e19af17dfd399b5626073d5f2a9
  - this simplies code and saves a lock for each call
- Skip if pod does not have claims: https://github.com/kubernetes/kubernetes/pull/72953/commits/c2d25e08d79c071c3b8405e98c71251a6dc6fce4
  - most pods does not use claims, there is no need to find pod volumes for them (suggested by @msau42 before in an [old PR](https://github.com/kubernetes/kubernetes/pull/71212#discussion_r235111911) to clear stale pod binding cache too)
  - one thing I am not sure is should we skip `VolumeSchedulingStageLatency` metric too?
- Clear cache instead of saving nils if no claims to bind or provision: https://github.com/kubernetes/kubernetes/pull/72953/commits/dbd80460de04fd1097065dee646b667c0468f2fd
  - This saves memory.

Why: https://github.com/kubernetes/kubernetes/pull/72045#issuecomment-454602165

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
/sig storage
/sig scheduling

### CODE REVIEW

File: pkg/controller/volume/persistentvolume/scheduler_binder.go [hunk 1]
```diff
@@ -142,10 +142,40 @@ func (b *volumeBinder) GetBindingsCache() PodBindingCache {
...
+
+	// Warning: Below log needs high verbosity as it can be printed several times (#60933).
+	klog.V(5).Infof("FindPodVolumes for pod %q, node %q", podName, node.Name)
+
+	// Initialize to true for pods that don't have volumes
+	unboundVolumesSatisfied = true
+	boundVolumesSatisfied = true
+	start := time.Now()
+	defer func() {
+		VolumeSchedulingStageLatency.WithLabelValues("predicate").Observe(time.Since(start).Seconds())
```
Comment (cofyc) [inline pkg/controller/volume/persistentvolume/scheduler_binder.go]: one thing I am not sure is should we skip VolumeSchedulingStageLatency metric too?

File: pkg/controller/volume/persistentvolume/scheduler_binder_test.go [hunk 2]
```diff
@@ -328,23 +328,19 @@ func (env *testEnv) assumeVolumes(t *testing.T, name, node string, pod *v1.Pod,
...
-	env.internalBinder.podBindingCache.UpdateProvisionedPVCs(pod, node, provisionings)
+	env.internalBinder.podBindingCache.UpdateBindings(pod, node, bindings, provisionings)
 }
 
 func (env *testEnv) initPodCache(pod *v1.Pod, node string, bindings []*bindingInfo, provisionings []*v1.PersistentVolumeClaim) {
 	cache := env.internalBinder.podBindingCache
-	cache.UpdateBindings(pod, node, bindings)
-
-	cache.UpdateProvisionedPVCs(pod, node, provisionings)
+	cache.UpdateBindings(pod, node, bindings, provisionings)
```
Comment (krmayankk) [inline pkg/controller/volume/persistentvolume/scheduler_binder_test.go]: do we need to add UT for UpdateBindings with provisionings param nil or few different values of it ?

Comment (cofyc) [reply pkg/controller/volume/persistentvolume/scheduler_binder_test.go]: Sure, UpdateBindings/UpdateProvisionedPVCs has been tested in  [scheduler_binder_cache_test.go](https://github.com/kubernetes/kubernetes/blob/f4e4a69aefe53ca8cd4d3fc954beba7700c83d1a/pkg/controller/volume/persistentvolume/scheduler_binder_cache_test.go#L63), we can add more cases.

**Comment by cofyc** [comment]
Found a flaky test, but it's out of scope of this PR. I opened another issue https://github.com/kubernetes/kubernetes/issues/72954.

**Comment by cofyc** [comment]
FIY, overview of FindPodVolumes after this PR:

- a) Pod has no PVCs, skip finding volumes and return without error
   - before: save nils in pod binding cache
- b) Pod has PVCs
  - 1) successfully found some claims to bind/provision, save them in pod binding cache
  - 2) successfully found no claims to bind/provision, clear cache in pod binding cache
     - before: save nils in pod binding cache
  - 3) failed to find claims to bind/provision, clear cache in pod binding cache

File: pkg/controller/volume/persistentvolume/scheduler_binder.go (hunk 1 above)
Comment (msau42) [reply pkg/controller/volume/persistentvolume/scheduler_binder.go]: Let's still keep it. It's good to measure when we don't do anything too.

**Comment by bsalamat** [comment]
Early benchmark results show that the regression caused by #72045 is fixed after this PR.
Thanks, @cofyc!

**Pull request merged**